### PR TITLE
🤯 Delete implementation

### PR DESCRIPTION
**Pull request by ccamel**
This PR brings addresses #307 by implementing the (missing) DeleteData message in the Cognitarium smart contract, which provides the functionality to delete triples from the contract's state.

:warning: I had to change the contract of the `DELETE` message so that the delete triple patterns are now optional but the where clause is compulsory. This change aligns with the W3C's [SPARQL DELETE query](https://www.w3.org/TR/sparql11-update/#deleteWhere) specification, which serves as our reference point.

### CODE REVIEW

**Comment by ccamel** [comment]
Status: implementation done. Missing tests.

**Comment by ccamel** [comment]
Rebased to include #311 and #322 (among others).

**Review by amimart — CHANGES_REQUESTED** [review]

File: contracts/okp4-cognitarium/src/storer/engine.rs [hunk 1]
```diff
@@ -91,18 +91,48 @@ impl<'a> TripleStorer<'a> {
...
+            self.delete_triple(atom)?;
+        }
+        self.finish()
+    }
+
+    pub fn delete_triple(&mut self, atom: &rdf::Atom) -> Result<(), ContractError> {
+        let triple = self.rio_to_triple(atom.into())?;
+        let object_hash: Hash = triple.object.as_hash();
+
+        self.store.stat.triple_count -= Uint128::one();
```
[the remark this answers is not in the document]
Comment (ccamel) [reply contracts/okp4-cognitarium/src/storer/engine.rs:105]: That's a good catch. Thanks :relieved:

**Review by amimart — APPROVED** [review]
Looks good! 🚀

**Review by bdeneux — APPROVED** [review]
Seems good 👍 Nice !

**Pull request merged**